src/com/opc/opceventhandler.cpp

```diff
@@ -19,16 +22,39 @@
 
 DEFINE_HANDLER(COpcEventHandler);
 
+CSyncObject COpcEventHandler::mSync;
+forte::arch::CSemaphore COpcEventHandler::mStateSemaphore;
+bool COpcEventHandler::mIsSemaphoreEmpty = true;
+
 COpcEventHandler::TCallbackDescriptor COpcEventHandler::mCallbackDescCount = 0;
 
 COpcEventHandler::COpcEventHandler(CDeviceExecution& paDeviceExecution) : CExternalEventHandler(paDeviceExecution)  {
-  this->start();
-  // Sleep to allow new thread to start
-  CThread::sleepThread(100);
+  if (!isAlive()) {
+    this->start();
+    // Sleep to allow new thread to start
+    CThread::sleepThread(100);
+  }
 }
 
 COpcEventHandler::~COpcEventHandler(){
-  this->end();
+  if (isAlive()) {
+    setAlive(false);
+    resumeSelfSuspend();
+    end();
+  }
+
+  clearCommandQueue(); // delete command in queue if exist
+}
+
+void COpcEventHandler::clearCommandQueue(){
+  while(!mCommandQueue.isEmpty()) {
+    // should never have come in here
```

Comment:
@azoitl Sir, with a "normal" disableHandler routine, there are always some commands left in the queue on kill. I write this clearCommandQueue actually to debug what command left there.